### PR TITLE
982 error this iri is already being processed

### LINKED ISSUE #982: Error "This iri <...> is already being processed"
Ontohub flags https://ontohub.org/cl-structural-ontology/distributive.clif 
as "processing failed", but I cannot figure out why. My local Hets version parses it fine. 

Is there a way to see the error message? If not, then this is a feature request. :-) 
Ideally the errors symbol would link to some detailed information.

### PR DESCRIPTION
**Pull request by eugenk**
This is mostly a refactoring. It shall finally fix #982. It
* introduces a more sophisticated Semaphore that can also lock and unlock in separate methods.
* splits the Redis namespaces between the environments (production, development, test)
* shares the Redis connection between Sidekiq and the Semaphore. Other services can easily be included to use the same connection as well.
* all `Worker`s are now implemented as intended by Sidekiq
* changes completely the handling of parsing multiple ontologies in parallel: We don't have many complex workers that handle parallelity themselves. Instead, we use the Semaphore on the ontology Loc/Id

I will not merge this before #1618 is merged, because both change some things with the `Worker` classes and it should be way easier to rebase this on #1618 + staging.

Things to do:
* [ ] merge #1618
* [ ] rebase on staging after that merge
* [x] merge this into develop and deploy on develop.ontohub.org.

### CODE REVIEW

File: lib/ontology_parsing_worker.rb [hunk 1]
```diff
@@ -1,20 +1,63 @@
...
+  # We cannot use Sidekiq's retry feature here because we need to retry per
+  # OntologyVersion in the queue of OntologyVersions
+  sidekiq_options retry: false
+
+  # queue is an array of triples like
+  # [ontology_version_id,
+  #  {fast_parse: true/false, files_to_parse_afterwards: array of strings},
+  #  try_count]
+  # where try_count is the number of the next evaluation attempt
+  def perform(version_with_options_queue)
```
Comment (eugenk) [inline lib/ontology_parsing_worker.rb]: I don't think we can shorten this reasonably.

File: lib/hets/concurrent_evaluator.rb [hunk 2]
```diff
@@ -5,18 +5,23 @@ module Hets
...
+    EVALUATION_MUTEX_KEY_PREFIX = "#{self}-evaluating:"
+    EVALUATION_MUTEX_EXPIRATION = HetsInstance::FORCE_FREE_WAITING_PERIOD
+
+    delegate :semaphore_stack, to: :importer
     delegate :ontologies_count, to: :importer
 
     def process(node_type, order, *args)
       super(node_type, order, *args)
-    rescue Exception => e
+    rescue Exception
```
Comment (eugenk) [inline lib/hets/concurrent_evaluator.rb]: We immediately re-raise, so this is okay.

File: lib/semaphore.rb [hunk 3]
```diff
@@ -5,20 +5,97 @@
...
+# after `expiration` seconds.
+#
+# It takes a block containing the critical path.
 class Semaphore
-  def self.exclusively(lock_key, &block)
-    if defined?(Sidekiq::Testing) && Sidekiq::Testing.inline?
-      yield
-    else
+  SEMAPHORES_NAMESPACE =
+    "#{Settings.redis.namespace}:#{self.to_s.downcase}".freeze
```
Comment (eugenk) [inline lib/semaphore.rb]: I'd rather leave this one in there to make sure one understands that this is the class name

**Comment by ebolloff** [comment]
I tested it like this: Shut down Sidekiq, create a fork from a repo, add a URL map and start sidekiq again. With your test repo `test-import-dol` everything is fine but with `test-import-clif` I still get this error with all ontologies `failed: target ontology not found: /iri2/o1//o1`.

**Comment by eugenk** [comment]
Then this is an error of the ontology and/or ontology parsing. It's not related to this pull request. Please try mirroring a repository with some imports on develop.ontohub.org instead of doing it locally. There, we have true parallelity.

**Comment by ebolloff** [comment]
:+1:

**Comment by eugenk** [comment]
The commits are all "new" because I rebased on staging. There were no conflicts. Merging now.

**Comment by eugenk** [comment]
The commits are all "new" because I rebased on staging after merging #1618. There were conflicts, because two worker files that were corrected here were deleted on staging. They also should be deleted because the whole parsing worker structure has changed.

These were the files:
* spec/lib/ontology_batch_parse_worker_spec.rb
* spec/lib/worker_spec.rb.

**Comment by eugenk** [comment]
About failing tests: I think we can ignore jenkins for now because it is being restructured anyway. Travis and development-machine-local tests should be green, though.

**Pull request merged**